### PR TITLE
Completed MCScript ETL. Working on PhysicalIQA

### PR DESCRIPTION
**Pull request by forbug**
I think we are supposed to create a new branch before doing this, but I wanted to get my code up in case you wanted to take a look. Let me know if I should fix anything going forward!

### CODE REVIEW

File: mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from mcs_benchmark_data.models.benchmark_test_dataset import BenchmarkTestDataset
+from mcs_benchmark_data.models.benchmark_dev_dataset import BenchmarkDevDataset
+from mcs_benchmark_data.models.benchmark_goal import BenchmarkGoal
+from mcs_benchmark_data.models.benchmark_question_type import BenchmarkQuestionType
+from mcs_benchmark_data.models.benchmark_sample import BenchmarkSample
+
+
+class PhysicalIQaBenchmarkTransformer(_Transformer):
+
+    __URI_BASE = "benchmark:physicaliqa"
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py:27]: Duplicating this.

How about creating an ABC _BenchmarkTransformer that has
```
@property
def _uri_base(self):
   return f"benchmark:{self._pipeline_id}"
```
or something like that.

File: mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        "dev": BenchmarkDevDataset,
+        "test": BenchmarkTestDataset,
+        "train": BenchmarkTrainDataset,
+    }
+
+    def transform(
+        self, benchmark_json_file_path: Path, **kwds
+    ) -> Generator[_Model, None, None]:
+
+        with open(benchmark_json_file_path) as benchmark_json:
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py:38]: Duplicating this code again. Put all this in the ABC?

File: mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        benchmark = Benchmark(
+            uri=URIRef(f"{self.__URI_BASE}:benchmark:{benchmark_metadata['@id']}"),
+            name=benchmark_bootstrap.name,
+            abstract=benchmark_bootstrap.abstract,
+            authors=tuple(author["name"] for author in benchmark_bootstrap.authors),
+        )
+
+        yield benchmark
+
+        for dataset in benchmark_metadata["datasets"]:
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py:52]: More duplicated code. Calldowns from the ABC?

File: mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+            goal = BenchmarkGoal(
+                uri=URIRef(f"{benchmark_sample.uri}:goal"),
+                benchmark_sample_uri=benchmark_sample.uri,
+                text=sample["goal"],
+            )
+
+            yield goal
+
+            sol1 = BenchmarkHypothesis(
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_transformer.py:128]: If you're not going to use a variable you assign something to, you can just do `yield BenchmarkHypothesis(...`.

File: tests/mcs_benchmark_data_test/pipelines/mcscript/mcscript_benchmark_extractor_test.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from mcs_benchmark_data.models.benchmark import Benchmark
+from mcs_benchmark_data.models.benchmark_dataset import BenchmarkDataset
+from mcs_benchmark_data.pipelines.mcscript.mcscript_benchmark_extractor import (
+    MCScriptBenchmarkExtractor,
+)
+
+
+def test_extract():
+    MCScriptBenchmarkExtractor(
```
Comment (gordom6) [inline tests/mcs_benchmark_data_test/pipelines/mcscript/mcscript_benchmark_extractor_test.py:9]: Need assertions.

File: tests/mcs_benchmark_data_test/pipelines/mcscript/mcscript_pipeline_test.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+from rdflib.compare import to_isomorphic
+from rdflib.parser import StringInputSource
+
+from mcs_benchmark_data.path import DATA_DIR_PATH
+from mcs_benchmark_data.pipelines.mcscript.mcscript_benchmark_pipeline import (
+    MCScriptBenchmarkPipeline,
+)
+
+
+def test_extract_transform_load(tmp_path):
```
Comment (gordom6) [inline tests/mcs_benchmark_data_test/pipelines/mcscript/mcscript_pipeline_test.py:14]: You're not using tmp_path.

**Review by gordom6 — CHANGES_REQUESTED** [review]

File: mcs_benchmark_data/_benchmark_file_names.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+from typing import NamedTuple
+from dataclasses import dataclass
+from dataclasses_json import LetterCase, dataclass_json
+
+
+@dataclass_json(letter_case=LetterCase.CAMEL)
+@dataclass(frozen=True)
+class _BenchmarkFileNames:
+    meta_data: str = "metadata.json"
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_file_names.py:9]: metadata_json

File: mcs_benchmark_data/_benchmark_submission_transformer.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+class _BenchmarkSubmissionTransformer(_Transformer):
+
+    """
+    Abstract base class for benchmark transformers.
+    See the transform method of _Transformer.
+    """
+
+    @property
+    def _uri_base(self):
+        return f"benchmark:{self._pipeline_id}"
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_submission_transformer.py:25]: benchmark:submission:?

File: mcs_benchmark_data/_benchmark_submission_transformer.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+from mcs_benchmark_data.models.submission import Submission
+from mcs_benchmark_data.models.submission_sample import SubmissionSample
+from mcs_benchmark_data.models.test_score import TestScore
+from mcs_benchmark_data.models.dev_score import DevScore
+
+
+class _BenchmarkSubmissionTransformer(_Transformer):
+
+    """
+    Abstract base class for benchmark transformers.
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_submission_transformer.py:19]: Benchmark submissions?

Comment (gordom6) [reply mcs_benchmark_data/_benchmark_submission_transformer.py:19]: I assume you're not using this class yet, probably should have waited for another PR, leave it for now.

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+from mcs_benchmark_data.models.benchmark import Benchmark
+from mcs_benchmark_data.models.benchmark_bootstrap import BenchmarkBootstrap
+from mcs_benchmark_data.models.benchmark_sample import BenchmarkSample
+from mcs_benchmark_data.models.benchmark_question_type import BenchmarkQuestionType
+from mcs_benchmark_data.models.benchmark_train_dataset import BenchmarkTrainDataset
+from mcs_benchmark_data.models.benchmark_test_dataset import BenchmarkTestDataset
+from mcs_benchmark_data.models.benchmark_dev_dataset import BenchmarkDevDataset
+
+
+class _Benchmark_Transformer(_Transformer):
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:21]: _BenchmarkTransformer

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    Abstract base class for transformers.
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:23]: benchmark transformers

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    def _transform(self, **kwds) -> Generator[_Model, None, None]:
+
+        benchmark_json_file_path = kwds["extracted_path"] / getattr(
+            kwds["file_names"], "meta_data"
+        )
+
+        with open(benchmark_json_file_path) as benchmark_json:
+            benchmark_metadata = json.loads(benchmark_json.read())
+
+        benchmark_bootstrap = BenchmarkBootstrap.from_dict(benchmark_metadata)
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:48]: Might call this dataclass BenchmarkMetadata instead of BenchmarkBootstrap, so the names line up. Fine to leave for another PR.

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+        yield benchmark
+
+        for dataset in benchmark_metadata["datasets"]:
+
+            dataset_type = dataset["@id"].split("/")[-1]
+
+            dataset_uri = URIRef(f"{self._uri_base}:dataset:{dataset['@id']}")
+
+            new_dataset = self._benchmark_dataset_classes[dataset_type](
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:65]: If you're only using this dict in this one place, it doesn't need to be protected. Can be a private class member.

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        self,
+        *,
+        dataset_type: str,
+        dataset_uri: URIRef,
+        sample: dict,
+        correct_choice: int,
+        **kwds,
+    ) -> Generator[_Model, None, None]:
+
+        correct_choice = None
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:87]: Immediately re-assign a parameter?

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            name=benchmark_bootstrap.name,
+            abstract=benchmark_bootstrap.abstract,
+            authors=tuple(author["name"] for author in benchmark_bootstrap.authors),
+        )
+
+        yield benchmark
+
+        for dataset in benchmark_metadata["datasets"]:
+
+            dataset_type = dataset["@id"].split("/")[-1]
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:61]: Should probably be a string enum and not a str. See the Python enum module.

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+            yield from self._transform_benchmark_sample(
+                dataset_type=dataset_type,
+                dataset_uri=new_dataset.uri,
+                **kwds,
+            )
+
+    def _prepare_sample(
+        self,
+        *,
+        dataset_type: str,
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:80]: dataset_type isn't used anywhere, so you don't need to declare it, can just be absorbed into **kwds if you're not using it.

File: mcs_benchmark_data/_benchmark_transformer.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+            yield new_dataset
+
+            yield from self._transform_benchmark_sample(
+                dataset_type=dataset_type,
+                dataset_uri=new_dataset.uri,
+                **kwds,
+            )
+
+    def _prepare_sample(
```
Comment (gordom6) [inline mcs_benchmark_data/_benchmark_transformer.py:77]: Consistent naming, please. _prepare_benchmark_sample.

File: mcs_benchmark_data/loaders/rdf_file_loader.py [hunk 18]
```diff
@@ -26,7 +26,7 @@ def __init__(
         *,
         compress: bool = True,
         file_path: Optional[Path] = None,
-        format="json-ld",
+        format="turtle",
```
Comment (gordom6) [inline mcs_benchmark_data/loaders/rdf_file_loader.py:29]: ttl

Also, please don't check in the .ttl file, only the .ttl.bz2 produced by the loader.

File: mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_extractor.py [hunk 19]
```diff
@@ -2,35 +2,23 @@
...
+
 
 class CommonsenseQaBenchmarkExtractor(_Extractor):
     def __init__(
         self,
-        dev_jsonl_file_name: str,
-        test_jsonl_file_name: str,
-        train_jsonl_file_name: str,
-        **kwds
+        file_names: CommonsenseQaBenchmarkFileNames,
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_extractor.py:13]: With this change to file_names, you don't really need benchmark-specific extractors any more, since they'll all look the same (take file_names, return it via extract). In effect you've moved the benchmark specialization to the *FileNames class hierarchy. You could call this class something like FileBenchmarkExtractor, or maybe just BenchmarkExtractor, since we don't have any non-file cases that I'm aware of.

File: mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_transformer.py [hunk 20]
```diff
@@ -7,120 +7,70 @@
...
-from mcs_benchmark_data.models.benchmark_dev_dataset import BenchmarkDevDataset
 from mcs_benchmark_data.models.benchmark_question import BenchmarkQuestion
 from mcs_benchmark_data.models.benchmark_question_type import BenchmarkQuestionType
 from mcs_benchmark_data.models.benchmark_sample import BenchmarkSample
 
 
-class CommonsenseQaBenchmarkTransformer(_Transformer):
+class CommonsenseQaBenchmarkTransformer(_Benchmark_Transformer):
+    def transform(self, **kwds) -> Generator[_Model, None, None]:
+        yield from _Benchmark_Transformer._transform(self, **kwds)
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_transformer.py:23]: Looks good.

File: mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_transformer.py [hunk 21]
```diff
@@ -7,120 +7,70 @@
...
-
-        benchmark_bootstrap = BenchmarkBootstrap.from_dict(benchmark_metadata)
-
-        benchmark = Benchmark(
-            uri=URIRef(f"{self.__URI_BASE}:benchmark:{benchmark_metadata['@id']}"),
-            name=benchmark_bootstrap.name,
-            abstract=benchmark_bootstrap.abstract,
-            authors=tuple(author["name"] for author in benchmark_bootstrap.authors),
+        sample_jsonl_file_path = kwds["extracted_path"] / getattr(
+            kwds["file_names"], dataset_type + "_samples"
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/commonsense_qa/commonsense_qa_benchmark_transformer.py:30]: Declare file_names rather than reaching into the kwds dict like this.

File: mcs_benchmark_data/_benchmark_transformer.py (hunk 17 above)
Comment (gordom6) [reply mcs_benchmark_data/_benchmark_transformer.py:77]: Seems like this is also specialized toward multiple choice, as well, and it's really a transform method. _transform_multiple_choice_benchmark_sample?

Comment (gordom6) [reply mcs_benchmark_data/_benchmark_transformer.py:77]: Looking further down, I think you prematurely factored this out, and it's not really reusable code. For example, you have to recreate the sample URI in your subclass, so you have duplicate code for that.

There's so little code here that separating it into the parent class obfuscates what's going on.

File: mcs_benchmark_data/pipelines/commonsense_qa/kagnet_commonsense_qa_submission_transformer.py [hunk 22]
```diff
@@ -1,28 +1,20 @@
...
+    _BenchmarkSubmissionTransformer,
 )
-from mcs_benchmark_data.models.submission import Submission
-from mcs_benchmark_data.models.submission_sample import SubmissionSample
-from mcs_benchmark_data.models.test_score import TestScore
-from mcs_benchmark_data.models.dev_score import DevScore
 
 
-class KagnetCommonsenseQaSubmissionTransformer(_CommonsenseQaSubmissionTransformer):
+class KagnetCommonsenseQaSubmissionTransformer(_BenchmarkSubmissionTransformer):
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/commonsense_qa/kagnet_commonsense_qa_submission_transformer.py:12]: I'd have preferred leaving the submission-related changes to a different PR.

File: mcs_benchmark_data/pipelines/commonsense_qa/roberta_commonsense_qa_submission_transformer.py [hunk 23]
```diff
@@ -30,75 +21,8 @@ def transform(
...
 
-        # Yield submissions
-        # Assumes file name in form "*_[systemname]_submission.jsonl" (e.g. dev_rand_split_roberta_submission.jsonl)
-        yield from self.__transform_submission(submission_data_jsonl_file_path)
-
-        submission_uri = URIRef(f"{self.__URI_BASE}:submission:CommonsenseQA-roberta")
-
-        yield from self.__transform_submission_sample(
-            submission_jsonl_file_path, submission_uri
+        yield from self._transform(
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/commonsense_qa/roberta_commonsense_qa_submission_transformer.py:24]: I don't think this is really reusable code, either, since it's specialized toward jsonl? But we can address that in a different PR.

File: mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_pipeline.py [hunk 24]
```diff
@@ -6,29 +6,20 @@
...
-    def __init__(
-        self,
-        dev_jsonl_file_name="dev.jsonl",
-        dev_labels_file_name="dev-labels.lst",
-        test_jsonl_file_name="test.jsonl",
-        train_jsonl_file_name="train.jsonl",
-        train_labels_file_name="train-labels.lst",
-        **kwds
-    ):
+    def __init__(self, file_names=PhysicalIQaBenchmarkFileNames, **kwds):
```
Comment (gordom6) [inline mcs_benchmark_data/pipelines/physical_iqa/physical_iqa_benchmark_pipeline.py:17]: Should be an instance, not the class?

**Review by gordom6 — CHANGES_REQUESTED** [review]

File: mcs_benchmark_data/_benchmark_file_names.py (hunk 7 above)
Comment (gordom6) [reply mcs_benchmark_data/_benchmark_file_names.py:9]: This should just be `metadata`, fitting with the rest of your naming schema.

File: mcs_benchmark_data/_benchmark_transformer.py (hunk 16 above)
Comment (forbug) [reply mcs_benchmark_data/_benchmark_transformer.py:80]: It is used in this function in line 88.

File: mcs_benchmark_data/_benchmark_transformer.py (hunk 15 above)
Comment (forbug) [reply mcs_benchmark_data/_benchmark_transformer.py:61]: What are you referring to? The benchmark_bootstrap model?

File: mcs_benchmark_data/_benchmark_transformer.py (hunk 14 above)
Comment (forbug) [reply mcs_benchmark_data/_benchmark_transformer.py:87]: I must have been trained about scope incorrectly. I believed that if I didn't assign it to None here, then I would receive an error if a test sample tried to use it in its declaration before being assigned. Is this not true?

File: mcs_benchmark_data/pipelines/commonsense_qa/roberta_commonsense_qa_submission_transformer.py (hunk 23 above)
Comment (forbug) [reply mcs_benchmark_data/pipelines/commonsense_qa/roberta_commonsense_qa_submission_transformer.py]: Ok I will check in before I move anything around, because now I am uncertain what code you actually wanted me to put into the parent class.

File: .gitignore [hunk 25]
```diff
@@ -1,4 +1,5 @@
 /env/
 __pycache__/
 /venv/
-
+mcs-benchmark-data/data/loaded/CommonsenseQA/CommonsenseQA.ttl
```
Comment (gordom6) [inline .gitignore:4]: Paths are relative to the .gitignore.

**Review by gordom6 — APPROVED** [review]
